program/src/oracle/upd_aggregate.h

```diff
@@ -313,38 +325,58 @@ static inline void upd_aggregate( pc_price_t *ptr, uint64_t slot )
   if ( rdenom->v_ ) {
     pd_div( rnumer, rnumer, rdenom );
   }
-  pd_new( cumwgt, 0, 0 );
-  pd_new( half, 5, -1 );
-  for( uint32_t i=0;i != numa; ++i ) {
-    wptr = &qs->weight_[i];
-    if ( aidx[i] == 0 ) {
+  for ( uint32_t i = 0; i != numa; ++i ) {
+    wptr = &qs->weight_[ i ];
+    if ( aidx[ i ] == 0 ) {
       pd_mul( wptr, wptr, rnumer );
     }
-    pd_mul( weight, wptr, half );
-    pd_add( &qs->cumwgt_[i], cumwgt, weight, qs->fact_ );
-    pd_add( cumwgt, cumwgt, wptr, qs->fact_ );
   }
 
+  const pd_t half = { .e_ = -1, .v_ = 5 };
+
   // compute aggregate price as weighted median
   pd_t iprice[1], lprice[1], uprice[1], q3price[1], q1price[1], ptile[1];
   pd_new( ptile, 5, -1 );
-  wgt_ptile( iprice, qs->iprice_, ptile, qs );
+  wgt_ptile( iprice, qs->iprice_, qs->weight_, numa, ptile, qs );
   pd_adjust( iprice, ptr->expo_, qs->fact_ );
   ptr->agg_.price_  = iprice->v_;
 
   // compute diff in weighted median between upper and lower price bounds
-  wgt_ptile( uprice, qs->uprice_, ptile, qs );
-  wgt_ptile( lprice, qs->lprice_, ptile, qs );
+  pd_t prices[ PC_COMP_SIZE ];
+  pd_t weights[ PC_COMP_SIZE ];
+  // sort upper prices and weights
+  for ( uint32_t i = 0; i < numa; ++i ) {
+    uint32_t j = i;
+    for ( ; j > 0 && pd_lt( &qs->uprice_[ i ], &prices[ j - 1 ], qs->fact_ ); --j ) {
```

Comment:
I wonder if using the qsort you added would be more efficient w.r.t instructions (add all, then qsort once)

Reply:
Thanks - started down that route but it added complexity (we have to move weight values in parallel), and the `uprice_` is already mostly sorted, so decided to use insertion sort for now.